File: c++/c06_4_minimizing_sums.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	double min1(vector<double> ary) {
		if (ary.empty()) return 0;
		ll idx = ssize(ary) / 2;
		ranges::nth_element(ary, ary.begin() + idx);
		return ary[idx];
	}

	double sum1(const vector<double>& ary, double x) {
		return ranges::fold_left(ary | views::transform([x](double d) { return abs(d - x); }), 0, plus());
	}

	double min2(const vector<double>& ary) {
		return ary.empty() ? 0 : ranges::fold_left(ary, 0, plus()) / ary.size();
	}

	double sum2(const vector<double>& ary, double x) {
		return ranges::fold_left(ary | views::transform([x](double d) { return (d - x) * (d - x); }), 0, plus());
	}
}

TEST(C064MinimizingSums, min1sum1) {
	EXPECT_DOUBLE_EQ(sum1({}, min1({})), 0);
	EXPECT_DOUBLE_EQ(sum1({ 1, 2 }, min1({ 1, 2 })), 1);
	EXPECT_DOUBLE_EQ(sum1({ 1, 2, 9, 2, 6 }, min1({ 1, 2, 9, 2, 6 })), 12);
}

TEST(C064MinimizingSums, min2sum2) {
	EXPECT_DOUBLE_EQ(sum2({}, min2({})), 0);
	EXPECT_DOUBLE_EQ(sum2({ 1, 2, 9, 2, 6 }, min2({ 1, 2, 9, 2, 6 })), 46);
}

File: python/c06_4_minimizing_sums.py
import unittest
from c24_5_randomized_algorithms import order_statistics

def min1(ary):
    return order_statistics(ary,len(ary)//2) if ary else 0

def sum1(ary,x):
    return sum(abs(a-x) for a in ary)

def min2(ary):
    return sum(ary)/len(ary) if ary else 0

def sum2(ary,x):
    return sum((a-x)**2 for a in ary)

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(sum1([],min1([])),0)
        self.assertEqual(sum1([1,2],min1([1,2])),1)
        self.assertEqual(sum1([1,2,9,2,6],min1([1,2,9,2,6])),12)
        self.assertEqual(sum2([],min2([])),0)
        self.assertEqual(sum2([1,2,9,2,6],min2([1,2,9,2,6])),46)

if __name__=='__main__':
    unittest.main()
